README
==========
 ical2org
==========

ical2org is a command line tool for exporting data from the Mac OS X
application iCal so it can be used with the emacs tool `org-mode`_.
Data transfer is one-way only (from iCal to emacs), and is intended to
be used to show alarms and scheduled events managed by iCal within
org's agenda view.

Installation
============

Use normal Python application steps to install ical2org::

    $ pip install ical2org

or::

    $ easy_install ical2org

The source distribution is created using `Distribute`_, which will be
installed with ical2org.  The other dependencies (`pytz`_ and
`vobject`_) will also be installed automatically.

Generating org-mode Files
=========================

While ical2org supports multiple output formats, the default is an
org-mode outline.  Each calendar entry includes the summary, location,
date and time, and complete description (useful for searching).

For example, to extract the entries from a calendar called "Meetup:
Python", you would run::

    $ ical2org "Meetup: Python"
    Starting 14 days ago at 2010-01-03 00:00:00-05:00
    Ending 30 days from now at 2010-02-17 00:00:00-05:00
    Processing: Meetup: Python
      Python Atlanta February Meetup
    # -*- coding: utf-8 -*-
    * Meetup: Python	:clubs:
      :CATEGORY: Personal
    ** Python Atlanta February Meetup
       <2010-02-11 Thu 19:30-21:30>
       - PyAtl: Atlanta Python Programmers
         
         After an optional 6pm dinner at Figo Pasta at the corner of
         Howell Mill Road and Huff Road, we will meet as usual in the
         spacious and comfortable auditorium of the GTRI Food
         Processing Building at 7:30pm for our monthly meeting.
         
         Topics for this meeting:
         
         Brandon Rhodes will give his PyCon talk: "Learning Hosting
         Best-Practices From WebFaction"
         
         Atlanta, GA  30318 - USA
         
         Thursday, February 11 at 7:30 PM
         
         Photo: http://photos2.meetupstatic.com/photos/event/4/7/3/5/event_5238229.jpeg
         
         Attending: 3
         
         Details: http://www.meetup.com/python-atlanta/calendar/12325308/

Output is encoded as UTF-8, which may not work with Unix pipes and
shell redirection.  To save the output to a file instead of having it
printed to standard output, use the ``-o`` option.

By default, ical2org processes all of the calendars which are
displayed in iCal.  Hidden calendars are ignored unless you give their
name on the command line.

Generating Diary Files
======================

The default output format uses org's outline structure.  If you would
rather use emacs' diary-mode, use ``-f diary`` to switch modes.  The
diary entries do not include the full text of the event.

Recommended Use
===============

The recommended way to use ical2org is to export your iCal data at
least every day.  That gives you a moving window of fresh data, since
the date range for export is limited.  Save the output to a file
(e.g., ``~/Documents/Notes/calendar.org``) and add the file to org's
list of inputs for building agenda views.

If you use org-mode for taking notes in meetings, refile each event as
you make notes to avoid overwriting it the next time your calendars
are exported.

For example, you might run a command like::

    $ ical2org -o ~/Documents/Notes/calendar.org -b 30 -e 30

to export the data starting 30 days ago and ending 30 days from today
to the file ``calendar.org``.

Recurring Events
================

Recurring events in iCal are "exploded" to create separate entries in
the output.  This makes it easy to refile an event when you add notes,
without having the notes from one meeting occurance appear on the
wrong date.

Command Line Options
====================

-h                    show this help message and exit
-c CONFIG_FILENAME, --config=CONFIG_FILENAME
                      Configuration file name. Defaults to ~/.ical2org
-b DAYS_AGO, --begin=DAYS_AGO, --days-ago=DAYS_AGO
                      Number of days back in time to search. Defaults to 14.
-e DAYS_AHEAD, --end=DAYS_AHEAD, --days-ahead=DAYS_AHEAD
                      Number of days forward in time to search. Defaults to
                      30.
-v, --verbose         Increase verbose level
-q, --quiet           Turn off verbose output
-o OUTPUT_FILE_NAME, --output-file=OUTPUT_FILE_NAME
                      Write the output to the named file instead of stdout
--all                 Include all calendars, not just active.
--input-directory=INPUT_DIRECTORY
                      Directory containing calendars. Defaults to
                      ~/Library/Calendars.
-f FORMAT, --format=FORMAT
                      Output format. One of ['org', 'diary']. Defaults to
                      "diary".
--opt=OPT, --formatter-option=OPT
                      Formatter-specific option name=value
--help                Verbose help

.. _org-mode: http://orgmode.org/

.. _Distribute: http://packages.python.org/distribute/

.. _pytz: http://pytz.sourceforge.net/

.. _vobject: http://vobject.skyhouseconsulting.com/


debian/changelog
ical2org (1.0-1) unstable; urgency=low

  * Initial release.

 -- Karsten Heymann <karsten.heymann@gmx.de>  Mon, 18 Jan 2010 07:02:21 +0100


ical2org/app.py
# encoding: utf-8
#
# Copyright (c) 2009 Doug Hellmann.  All rights reserved.
#
"""Command line interface for ical2org.
"""

import codecs
from ConfigParser import SafeConfigParser as ConfigParser
import datetime
import inspect
import logging
import optparse
import os
import sys

from ical2org import calendars, filter, diary, tz, org

import vobject

VERBOSE_LEVELS = {
    0:logging.WARNING,
    1:logging.INFO,
    2:logging.DEBUG,
    }

FORMATTER_FACTORIES = {
    'diary':diary.DiaryFormatter,
    'org':org.OrgTreeFormatter,
    }

def remember_formatter_option(option, opt_str, value, parser):
    """Receive and handle a formatter-specific option.
    
    :param option: The parsed option results.
    :param opt_str: The option string from the command line.
    :param value: The option value.  Should be ``name=value``.
    :param parser: The option parser.
    """
    if not hasattr(option, 'formatter_options'):
        option.formatter_options = {}
    parsed = value.split('=', 1)
    if len(parsed) == 1:
        # flag
        option.formatter_options[value] = True
    else:
        # name=value
        opt_name, opt_val = parsed
        option.formatter_options[opt_name] = opt_val
    return

def show_verbose_help(option, opt_str, value, parser):
    parser.print_help()
    print '\nFormatters:\n'
    for name, formatter in sorted(FORMATTER_FACTORIES.items()):
        print name
        print '-' * len(name)
        print inspect.getdoc(formatter)
    parser.exit()
    return

def main(args=sys.argv[1:]):
    option_parser = optparse.OptionParser(
        usage='usage: ical2org [options] [calendar titles]',
        conflict_handler='resolve',
        description='Convert iCal calendar entries to org-mode data for use with emacs',
        )
    option_parser.add_option('-c', '--config',
                             action='store',
                             dest='config_filename',
                             help='Configuration file name. Defaults to ~/.ical2org',
                             default=os.path.expanduser('~/.ical2org'),
                             )
    option_parser.add_option('--begin', '-b', '--days-ago',
                             action='store',
                             dest='days_ago',
                             help='Number of days back in time to search. Defaults to 14.',
                             default=14,
                             type=int,
                             )
    option_parser.add_option('--end', '-e', '--days-ahead',
                             action='store',
                             dest='days_ahead',
                             help='Number of days forward in time to search. Defaults to 30.',
                             default=30,
                             type=int,
                             )
    option_parser.add_option('-v', '--verbose',
                             action='count',
                             dest='verbose_level',
                             default=1,
                             help='Increase verbose level',
                             )
    option_parser.add_option('-q', '--quiet',
                             action='store_const',
                             const=0,
                             dest='verbose_level',
                             help='Turn off verbose output',
                             )
    option_parser.add_option('--output-file', '-o',
                             action='store',
                             dest='output_file_name',
                             help='Write the output to the named file instead of stdout',
                             default=None,
                             )
    option_parser.add_option('--all',
                             action='store_false',
                             dest='active_only',
                             default=True,
                             help='Include all calendars, not just active.',
                             )
    option_parser.add_option('--input-directory',
                             action='store',
                             dest='input_directory',
                             default=os.path.expanduser('~/Library/Calendars'),
                             help='Directory containing calendars. Defaults to ~/Library/Calendars.',
                             )
    option_parser.add_option('--format', '-f',
                             action='store',
                             dest='format',
                             type='choice',
                             choices=FORMATTER_FACTORIES.keys(),
                             default='org',
                             help='Output format. One of %s. Defaults to "diary".' % FORMATTER_FACTORIES.keys(),
                             )
    option_parser.add_option('--opt', '--formatter-option',
                             action='callback',
                             type='string',
                             callback=remember_formatter_option,
                             help='Formatter-specific option name=value',
                             )
    option_parser.add_option('--help',
                             action='callback',
                             callback=show_verbose_help,
                             help='Verbose help',
                             )
                             
    options, calendar_titles = option_parser.parse_args(args)

    log_level = VERBOSE_LEVELS.get(options.verbose_level, logging.DEBUG)
    logging.basicConfig(level=log_level, format='%(message)s')

    config = ConfigParser()
    config.read([options.config_filename])

    # Compute the date range for items to be included in our output.
    start_date = tz.normalize_to_utc(datetime.datetime.combine(
        datetime.date.today() - datetime.timedelta(options.days_ago),
        datetime.time.min,
        ))
    end_date = tz.normalize_to_utc(datetime.datetime.combine(
        datetime.date.today() + datetime.timedelta(options.days_ahead + 1),
        datetime.time.min,
        ))
    logging.info('Starting %d days ago at %s', options.days_ago, start_date.astimezone(tz.local))
    logging.info('Ending %d days from now at %s', options.days_ahead, end_date.astimezone(tz.local))

    if options.output_file_name:
        logging.info('Writing to %s', options.output_file_name)

    # Load the calendars
    if calendar_titles:
        calendar_generator = calendars.get_by_titles(path=options.input_directory,
                                                     titles=calendar_titles)
    else:
        calendar_generator = calendars.discover(path=options.input_directory,
                                                active_only=options.active_only)

    # Process the calendar data
    output = sys.stdout
    if options.output_file_name:
        output = codecs.open(options.output_file_name, 'wt', 'UTF-8')
    try:
        formatter = FORMATTER_FACTORIES[options.format](output, config, options)
        for calendar in calendar_generator:
            logging.info('Processing: %s', calendar.title)
            formatter.start_calendar(calendar)
            for event in filter.unique(filter.by_date_range(calendar.get_events(),
                                                            start_date,
                                                            end_date,
                                                            )):
                logging.info('  %s', event.summary.value)
                formatter.add_event(event)
            formatter.end_calendar(calendar)
    finally:
        formatter.close()
        if output != sys.stdout:
            output.close()
    return
    
    
if __name__ == '__main__':
    main()
